rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -0,0 +1,689 @@
+parsing of a logical data format in a physical storage format is handled by a
+specific tf.Data Dataset that yields StructuredTensors.
+
+We also need to be able to construct a beam source from such a Dataset (although
+there is a non-trivial gap).
+
+The advantages of this approach are:
+
+*   Less effort to integrate: adding a new format == adding a new tf.Dataset.
+*   Unified story across TF and beam.
+*   No third-party dependencies.
+    *   Note that we can still provide an Arrow -> StructuredTensor adapter to
+        achieve interoperability.
+
+What this alternative does not change / address:
+
+*   We still need the TensorAdapter API to convert from StructuredTensor to
+    other TF feedables, unless we are willing to only offer StructuredTensors to
+    our end-users (which might be the case eventually, but not likely to happen
+    soon). So a good portion of this proposal will remain mostly unchanged.
+*   We still cannot expose file-based or record-based interface to end-users in
+    the TF trainer. As that is a direct result of I/O + parsing being abstracted
+    out.
+
+The disadvantages of this approach are:
+
+*   This tightly couples TF with TFX components, in the following ways:
+
+    *   Components will need TF to read the data in.
+    *   Components that analyze the data (e.g. TFDV) will operate against
+        StructuredTensors, and the easiest way to to conduct certain
+        computations (for example, slicing a StructuredTensor or computing the
+        mean) with StructuredTensors is through TF ops and their python
+        bindings.
+
+    Such a coupling is not in our favor because:
+
+    *   Some TFX components functionally do not require TF to work. For example,
+        TFDV can analyze any data set. TFMA can analyze a model while treating
+        the model as a blackbox. In both cases, the ML library that trained the
+        model is irrelevant and TF should not be assumed. And coupling with TF
+        does not only introduce a heavy dependency, but also forces the user to
+        learn about TF if they need to implement a TFXIO for their data format.
+
+    *   Operations against StructuredTensors boils down to Python + TF ops. The
+        overhead of either is much higher than just calling an Arrow C++ API
+        that does the same operation.
+
+    *   The main extension point of TF is Ops which don't understand the nested
+        structures. Compared to using Arrow’s C++ APIs, implementing an Op that
+        deals with StructuredTensors will be much complicated.
+
+# Questions and Discussion Topics
+
+## OSS build / release issues
+
+### We don’t have a strong representation in the Arrow community
+
+This has led to some issues. For example, the PyPI/Wheel packaging for pyarrow
+currently is unfunded and lacks volunteers and there is a risk of support being
+dropped, but we do rely on the pyarrow wheel as TFX is released on PyPI.
```

Comment:
This seems to argue that maybe the StructuredTensor approach is preferable? Can you comment on why this isn't an issue concerning enough to warrant not using Arrow?

Reply:
The odds of wheel support being dropped from the project seems extraordinarily unlikely. 

The pyarrow wheel has been installed almost 30 million times

```
$ pypistats overall pyarrow
|    category     | percent | downloads  |
|-----------------|--------:|-----------:|
| with_mirrors    |  50.16% | 14,307,951 |
| without_mirrors |  49.84% | 14,218,655 |
| Total           |         | 28,526,606 |
```

This is on the same order of magnitude as tensorflow and pandas

```
$ pypistats overall tensorflow
|    category     | percent | downloads  |
|-----------------|--------:|-----------:|
| with_mirrors    |  50.25% | 21,595,341 |
| without_mirrors |  49.75% | 21,377,666 |
| Total           |         | 42,973,007 |

$ pypistats overall pandas
|    category     | percent |  downloads  |
|-----------------|--------:|------------:|
| with_mirrors    |  50.10% |  86,117,641 |
| without_mirrors |  49.90% |  85,783,561 |
| Total           |         | 171,901,202 |
```

pyarrow is being depended on by some other Google products, like the BigQuery Python API (cc @tswast). 

My tiny not-for-profit org (https://ursalabs.org/) has been doing most of the wheel maintenance in 2019, perhaps all that is needed is for Google to allocate a little bit of engineering time to help with the wheels now and then?

Reply:
`pyarrow` seems to also be pulled in by `pip install apache_beam`

https://github.com/apache/beam/blob/master/sdks/python/setup.py#L122

Reply:
@brills - might be worth noting the above in the doc; as it is, the description sounds more dire than the apparent reality.

Reply:
thanks @wesm !

I removed that description about the risk of wheels being dropped.

Reply:
Yes we also use `pyarrow` in `apache_beam` (and thus Cloud Dataflow) for parquet support.